x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/elasticsearch/ElasticsearchInternalService.java

```diff
@@ -1004,4 +993,58 @@ static TaskType inferenceConfigToTaskType(InferenceConfig config) {
             return null;
         }
     }
+
+    // Iterates over the batch sending 1 request at a time to avoid
+    // filling the ml node inference queue.
+    class BatchIterator {
+        private final AtomicInteger index = new AtomicInteger();
+        private final ElasticsearchInternalModel esModel;
+        private final List<EmbeddingRequestChunker.BatchRequestAndListener> requestAndListeners;
+        private final InputType inputType;
+        private final TimeValue timeout;
+
+        BatchIterator(
+            ElasticsearchInternalModel esModel,
+            InputType inputType,
+            TimeValue timeout,
+            List<EmbeddingRequestChunker.BatchRequestAndListener> requestAndListeners
+        ) {
+            this.esModel = esModel;
+            this.requestAndListeners = requestAndListeners;
+            this.inputType = inputType;
+            this.timeout = timeout;
+        }
+
+        void run() {
+            inferenceExecutor.execute(() -> inferBatchAndRunAfter(requestAndListeners.get(index.get())));
+        }
+
+        private void inferBatchAndRunAfter(EmbeddingRequestChunker.BatchRequestAndListener batch) {
+            var inferenceRequest = buildInferenceRequest(
+                esModel.mlNodeDeploymentId(),
+                EmptyConfigUpdate.INSTANCE,
+                batch.batch().inputs(),
+                inputType,
+                timeout
+            );
+
+            ActionListener<InferModelAction.Response> mlResultsListener = batch.listener()
+                .delegateFailureAndWrap(
+                    (l, inferenceResult) -> translateToChunkedResult(esModel.getTaskType(), inferenceResult.getInferenceResults(), l)
+                );
+
+            // schedule the next request once the results have been processed
+            var runNextListener = ActionListener.runAfter(mlResultsListener, () -> {
+                if (index.incrementAndGet() < requestAndListeners.size()) {
+                    run();
+                }
+            });
+
+            var maybeDeployListener = runNextListener.delegateResponse(
+                (l, exception) -> maybeStartDeployment(esModel, exception, inferenceRequest, runNextListener)
```

Comment:
I'm a bit confused by this `l`. Shouldn't that appear on the rhs as well (instead of `runNextListener`)?

Reply:
In my experience you can use either listener. In this case, since the listener being passed is the same listener being delegated, it shouldn't make any difference which one you pass (although maybe there is some small hidden difference?). Sometimes you do need to pass a different listener than the delegate listener as well.

Reply:
functionally the same but very error prone to leave it as it is. I've replaced runNextListener with l